kvstore/pebble/pebble.go

```diff
@@ -0,0 +1,270 @@
+package pebble
+
+import (
+	"sync"
+
+	"github.com/cockroachdb/pebble"
+
+	"github.com/iotaledger/hive.go/byteutils"
+	"github.com/iotaledger/hive.go/kvstore"
+	"github.com/iotaledger/hive.go/types"
+)
+
+// peppleStore implements the KVStore interface around a pebble instance.
+type peppleStore struct {
+	instance                     *pebble.DB
+	dbPrefix                     []byte
+	accessCallback               kvstore.AccessCallback
+	accessCallbackCommandsFilter kvstore.Command
+}
+
+// New creates a new KVStore with the underlying pebbleDB.
+func New(db *pebble.DB) kvstore.KVStore {
+	return &peppleStore{
+		instance: db,
+	}
+}
+
+// AccessCallback configures the store to pass all requests to the KVStore to the given callback.
+// This can for example be used for debugging and to examine what the KVStore is doing.
+func (s *peppleStore) AccessCallback(callback kvstore.AccessCallback, commandsFilter ...kvstore.Command) {
+	var accessCallbackCommandsFilter kvstore.Command
+	if len(commandsFilter) == 0 {
+		accessCallbackCommandsFilter = kvstore.AllCommands
+	} else {
+		for _, filterCommand := range commandsFilter {
+			accessCallbackCommandsFilter |= filterCommand
+		}
+	}
+
+	s.accessCallback = callback
+	s.accessCallbackCommandsFilter = accessCallbackCommandsFilter
+}
+
+func (s *peppleStore) WithRealm(realm kvstore.Realm) kvstore.KVStore {
+	return &peppleStore{
+		instance: s.instance,
+		dbPrefix: realm,
+	}
+}
+
+func (s *peppleStore) Realm() []byte {
+	return s.dbPrefix
+}
+
+// builds a key usable for the pebble instance using the realm and the given prefix.
+func (s *peppleStore) buildKeyPrefix(prefix kvstore.KeyPrefix) kvstore.KeyPrefix {
+	return byteutils.ConcatBytes(s.dbPrefix, prefix)
+}
+
+// Shutdown marks the store as shutdown.
+func (s *peppleStore) Shutdown() {
+	if s.accessCallback != nil {
+		s.accessCallback(kvstore.ShutdownCommand)
+	}
+}
+
+func (s *peppleStore) Iterate(prefix kvstore.KeyPrefix, consumerFunc kvstore.IteratorKeyValueConsumerFunc) error {
+	if s.accessCallback != nil && s.accessCallbackCommandsFilter.HasBits(kvstore.IterateCommand) {
+		s.accessCallback(kvstore.IterateCommand, prefix)
+	}
+
+	start := s.buildKeyPrefix(prefix)
+	end := copyBytes(start)
+	end[len(end)-1] = end[len(end)-1] + 1
```

Comment:
This can crash with an empty realm. Also it will only work with realms, due to the range defined